### ISSUE TITLE
Support for changing the default compression level when enabling server response compression

### ISSUE THREAD

**Issue by juliojgd** [reactor/reactor-netty#3244]
## Motivation

In the [server compression related documentation](https://projectreactor.io/docs/netty/release/reference/index.html#_compression) the only configurable parameters are _enabled_ (boolean) and minimum size for a payload to be compressed (int).

There is [no method in Reactor Netty HttpServer](https://github.com/reactor/reactor-netty/blob/main/reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServer.java) that allows to set the compression level (int) that is effectively available in [Netty's HttpContentCompressor class](https://github.com/netty/netty/blob/4.1/codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java) but it's not available in the default [SimpleCompressionHandler](https://github.com/reactor/reactor-netty/blob/main/reactor-netty-http/src/main/java/reactor/netty/http/server/SimpleCompressionHandler.java)

It cannot be set in the [Reactor Netty HttpServerResponse](https://github.com/reactor/reactor-netty/blob/main/reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServerResponse.java) either.

## Desired solution

To add a configurable property (maybe named "compression-level" or "gzip-compression-level") to allow changing the default compression level (6 - gzip) to any other compression level that the developer needs to set.

**Comment by violetagg** [comment]
@juliojgd Are you interested in providing a PR? Which Reactor Netty version should enable this feature (1.1.x or 1.2.x)?

**Comment by juliojgd** [comment]
@violetagg we are using Spring Boot 3.2.x so I guess we'd need 1.1.x. I'll try to provide a PR.

The source branch to checkout for the changes would be https://github.com/reactor/reactor-netty/tree/1.1.x , right?

Thanks in advance

**Comment by violetagg** [comment]
> The source branch to checkout for the changes would be https://github.com/reactor/reactor-netty/tree/1.1.x , right?

yes

**Comment by raccoonback** [comment]
@violetagg 
Hello,
I would like to ask if I can take ownership of this issue and work on resolving it.
Additionally, please let me know if there are any guidelines or considerations I should be aware of before starting.

**Comment by violetagg** [comment]
@raccoonback Yes sure, go for it. Please follow the contribution guide https://github.com/reactor/.github/blob/main/CONTRIBUTING.md

**Comment by blacelle** [comment]
I encountered an issue which can be adressed by this feature. Thanks!

For any SpringBoot users, this can be wired with:

```
    @Bean
    NettyServerCustomizer nettyServerCustomizer() {
        return new NettyServerCustomizer() {
            @Override
            public HttpServer apply(HttpServer httpServer) {
                return httpServer.compressOptions(ZstdOption.builder().maxEncodeSize(64*1024*1024).build());
            }
        };
    }
```

One note if this feature does not enable disabling some compression schema (e.g. I see no way to prevent Zstd).

**Comment by violetagg** [comment]
@blacelle Reactor Netty doesn't have a dependency to Zstd. What do you mean that one cannot prevent Zstd?

**Comment by blacelle** [comment]
Naively, I would expect to be able to disable explicitely zstd amongst the available compression scheme. This seems not possible. I underatand it is an option in Netty,  which is not offered by Reactor-Netty (which enforced Zstd to be active if it is available).

---

I understand the compression scheme selection is done in `io.netty.handler.codec.http.HttpContentCompressor#determineEncoding`. This (Netty) method relies on pattern like:

```
if (zstdQ != -1.0F && zstdQ >= snappyQ && this.zstdOptions != null) {
    return "zstd";
}
```

suggesting `zstdOption` can be null (as a legitimate Netty configuration). `compressionOptions` (in netty) seems setup around:

```
    public HttpContentCompressor(int contentSizeThreshold, CompressionOptions... compressionOptions) {
        this.contentSizeThreshold = ObjectUtil.checkPositiveOrZero(contentSizeThreshold, "contentSizeThreshold");
        BrotliOptions brotliOptions = null;
        GzipOptions gzipOptions = null;
        DeflateOptions deflateOptions = null;
        ZstdOptions zstdOptions = null;
        SnappyOptions snappyOptions = null;
        if (compressionOptions == null || compressionOptions.length == 0) {
            compressionOptions = defaultCompressionOptions(StandardCompressionOptions.gzip(), StandardCompressionOptions.deflate());
        }
```

The input `compressionOptions` is crafted by Reactive_netty through `HttpCompressionOptionsSpec` (introduced in this ticket). However, I feel `HttpCompressionOptionsSpec` does not allow disabling such a scheme.

Specifically, I tried using:

```
    @Bean
    NettyServerCustomizer nettyServerCustomizer() {
        return new NettyServerCustomizer() {
            @Override
            public HttpServer apply(HttpServer httpServer) {
                return httpServer.compressOptions(GzipOption.builder().build());
            }
        };
    }
```
but `zstd` is not null by Reactive-Netty:

```
    private HttpCompressionOptionsSpec() {
        this.gzip = GzipOption.provideDefault();
        this.deflate = DeflateOption.provideDefault();
        this.snappy = new SnappyOption();
        if (Brotli.isAvailable()) {
            this.brotli = new BrotliOption();
        }

        if (Zstd.isAvailable()) {
            this.zstd = ZstdOption.provideDefault();
        }

    }
```

edit: `zstd` is just an example here, brought forward as there may be a Netty bug around it (https://github.com/netty/netty/issues/14972). But my point about scheme-deactivation is valid for any scheme (e.g. I may for any reason be willing to deactive gzip).

**Comment by violetagg** [comment]
@blacelle Basically we (Reactor Netty) support this for Brotli and Zstd. Reactor Netty doesn't have dependency to Brotli and Zstd so they will be null unless the user **explicitly** adds a dependency. In addition to that we need to support disabling for the rest. For Spring Boot and why you have dependency to Zstd, I cannot comment.
Would you like to provide a PR for this feature?

**Comment by blacelle** [comment]
Thanks @violetagg . Zstd is indeed a additional dependency in our case.

As a user, I'm happy with automated activation of Zstd/Brotli if available.
As a user, I would like to be able to explicitly (in Reactive-Netty configuration) disable any compression scheme (overriding any automated activation).

> Would you like to provide a PR for this feature?

I'm affraid I won't be able to open such a PR in the short-term (e.g. in profit of fixing the Zstd-Netty bug).

**Comment by violetagg** [comment]
@blacelle Would you like to open an enchantment request in case someone wants to picked it up?
